File: hexameter-core/src/test/java/org/codetome/hexameter/core/api/HexagonalGridBuilderTest.java
package org.codetome.hexameter.core.api;

import junit.framework.Assert;
import org.codetome.hexameter.core.api.exception.HexagonalGridCreationException;
import org.codetome.hexameter.core.internal.impl.layoutstrategy.GridLayoutStrategy;
import org.junit.Before;
import org.junit.Test;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNotNull;
import static org.codetome.hexameter.core.api.HexagonOrientation.FLAT_TOP;
import static org.codetome.hexameter.core.api.HexagonalGridLayout.RECTANGULAR;
import static org.codetome.hexameter.core.api.HexagonalGridLayout.TRIANGULAR;

public class HexagonalGridBuilderTest {

    private static final int GRID_HEIGHT = 9;
    private static final HexagonalGridLayout GRID_LAYOUT = RECTANGULAR;
    private static final GridLayoutStrategy GRID_LAYOUT_STRATEGY = RECTANGULAR.getGridLayoutStrategy();
    private static final int GRID_WIDTH = 9;
    private static final HexagonOrientation ORIENTATION = FLAT_TOP;
    private static final double RADIUS = 30;

    private HexagonalGridBuilder target;

    @Before
    public void setUp() {
        target = new HexagonalGridBuilder();
        target.setGridHeight(GRID_HEIGHT).setGridLayout(GRID_LAYOUT).setGridWidth(GRID_WIDTH).setOrientation(ORIENTATION).setRadius(RADIUS);
    }

    @Test
    public void shouldContainProperValuesWhenGettersAreCalled() {
        assertEquals(GRID_HEIGHT, target.getGridHeight());
        assertEquals(GRID_WIDTH, target.getGridWidth());
        assertEquals(GRID_LAYOUT_STRATEGY, target.getGridLayoutStrategy());
        assertEquals(RADIUS, target.getRadius());
        Assert.assertNotNull(target.getSharedHexagonData());
    }

    @Test(expected = IllegalStateException.class)
    public void shouldFailGettingSharedHexagonDataWhenOrientationIsNull() {
        target.setOrientation(null);
        target.getSharedHexagonData();
    }

    @Test(expected = IllegalStateException.class)
    public void shouldFailGettingSharedHexagonDataWhenRadiusIsZero() {
        target.setRadius(0);
        target.getSharedHexagonData();
    }

    @Test(expected = HexagonalGridCreationException.class)
    public void shouldFailBuildWhenOrientationIsNull() {
        target.setOrientation(null);
        target.build();
    }

    @Test(expected = HexagonalGridCreationException.class)
    public void shouldFailBuildWhenRadiusIsZero() {
        target.setRadius(0);
        target.build();
    }

    @Test(expected = HexagonalGridCreationException.class)
    public void shouldFailBuildWhenGridLayoutIsNull() {
        target.setGridLayout(null);
        target.build();
    }

    @Test(expected = HexagonalGridCreationException.class)
    public void shouldFailBuildWhenSizeIsNotCompatibleWithLayout() {
        target.setGridLayout(TRIANGULAR);
        target.setGridHeight(4);
        target.build();
    }

    @Test
    public void shouldBuildCalculatorWhenBuildCalculatorIsCalled() {
        final HexagonalGridCalculator calc = target.buildCalculatorFor(null);
        assertNotNull(calc);
    }

    @Test
    public void shouldReturnProperOrientationWhenGetOrientationIsCalled() {
        assertEquals(ORIENTATION, target.getOrientation());
    }

    @Test
    public void shouldBuildWhenProperParametersArePresent() {
        final HexagonalGrid grid = target.build();
        assertNotNull(grid);
    }

}

File: mixite2.core/src/commonTest/kotlin/org/hexworks/mixite2/core/api/HexagonalGridBuilderTest.kt
package org.hexworks.mixite2.core.api

import org.hexworks.mixite2.core.api.HexagonOrientation.FLAT_TOP
import org.hexworks.mixite2.core.api.HexagonalGridLayout.RECTANGULAR
import org.hexworks.mixite2.core.api.HexagonalGridLayout.TRIANGULAR
import org.hexworks.mixite2.core.api.contract.SatelliteData
import kotlin.test.*

class HexagonalGridBuilderTest {

    lateinit var target: HexagonalGridBuilder<out SatelliteData>

    @BeforeTest
    fun setUp() {
        target = HexagonalGridBuilder()
        target.setGridHeight(GRID_HEIGHT)
                .setGridLayout(GRID_LAYOUT)
                .setGridWidth(GRID_WIDTH)
                .setOrientation(ORIENTATION)
                .setRadius(RADIUS)
    }

    @Test
    fun shouldContainProperValuesWhenGettersAreCalled() {
        assertEquals(GRID_HEIGHT, target.getGridHeight())
        assertEquals(GRID_WIDTH, target.getGridWidth())
        assertEquals(GRID_LAYOUT_STRATEGY, target.gridLayoutStrategy)
        assertEquals(RADIUS, target.getRadius())
        assertNotNull(target.gridData)
    }

    @Test
    fun shouldFailGettingSharedHexagonDataWhenOrientationIsNull() {
        assertFailsWith<IllegalStateException> {
            target.setRadius(0.0)
            target.gridData
        }
    }


    @Test
    fun shouldFailBuildWhenSizeIsNotCompatibleWithLayout() {
        assertFailsWith<IllegalStateException> {
            target.setGridLayout(TRIANGULAR)
            target.setGridHeight(4)
            target.build()
        }
    }

    @Test
    fun shouldReturnProperOrientationWhenGetOrientationIsCalled() {
        assertEquals(ORIENTATION, target.getOrientation())
    }

    @Test
    fun shouldBuildWhenProperParametersArePresent() {
        val grid = target.build()
        assertNotNull(grid)
    }

    companion object {

        private const val GRID_HEIGHT = 9
        private const val GRID_WIDTH = 9
        private const val RADIUS = 30.0
        private val GRID_LAYOUT = RECTANGULAR
        private val GRID_LAYOUT_STRATEGY = RECTANGULAR.gridLayoutStrategy
        private val ORIENTATION = FLAT_TOP
    }

}
